src/fscacher/cache.py

```diff
@@ -119,8 +119,60 @@ def _get_file_fingerprint(path):
             # we can't take everything, since atime can change, etc.
             # So let's take some
             s = os.stat(path, follow_symlinks=True)
-            fprint = s.st_mtime_ns, s.st_ctime_ns, s.st_size
+            fprint = FileFingerprint.from_stat(s)
             lgr.log(5, "Fingerprint for %s: %s", path, fprint)
             return fprint
         except Exception as exc:
             lgr.debug(f"Cannot fingerprint {path}: {exc}")
+
+    @staticmethod
+    def _get_dir_fingerprint(path):
+        fprint = DirFingerprint()
+        dirqueue = deque([path])
+        try:
+            while dirqueue:
+                d = dirqueue.popleft()
+                with os.scandir(d) as entries:
+                    for e in entries:
+                        if e.is_dir(follow_symlinks=True):
+                            dirqueue.append(e.path)
+                        else:
+                            s = e.stat(follow_symlinks=True)
+                            fprint.add_file(e.path, FileFingerprint.from_stat(s))
+        except Exception as exc:
+            lgr.debug(f"Cannot fingerprint {path}: {exc}")
+            return None
+        else:
+            return fprint
+
+
+class FileFingerprint(namedtuple("FileFingerprint", "mtime_ns ctime_ns size")):
+    @classmethod
+    def from_stat(cls, s):
+        return cls(s.st_mtime_ns, s.st_ctime_ns, s.st_size)
+
+    def modified_in_window(self, min_dtime):
+        return abs(time.time() - self.mtime_ns * 1e-9) < min_dtime
+
+    def to_tuple(self):
+        return tuple(self)
+
+
+class DirFingerprint:
+    def __init__(self):
+        self.last_modified = None
+        self.tree_fprints = {}
+
+    def add_file(self, path, fprint: FileFingerprint):
+        self.tree_fprints[path] = fprint
+        if self.last_modified is None or self.last_modified < fprint.mtime_ns:
+            self.last_modified = fprint.mtime_ns
+
+    def modified_in_window(self, min_dtime):
+        if self.last_modified is None:
+            return False
+        else:
+            return abs(time.time() - self.last_modified * 1e-9) < min_dtime
+
+    def to_tuple(self):
+        return sum(sorted(self.tree_fprints.items()), ())
```

Comment:
`sorted` could likely be the "killer" of performance, while also would make us feel nice when we gain significant boost when we stop doing that ;)

would be important to provide some initial benchmarking setup, filed #12 with idea(s)

Reply:
Without `sorted`, the fingerprints would be unstable, and caching wouldn't work reliably.

Reply:
I understand the purpose of it, but I wonder if/how it could be avoided.  E.g. if "unstable" due to changes to the directory content -- until we introduce some include/exclude support, we should be fine even without it, which would help to lower the "expected caching overhead" and establish a new baseline to go from. If it is "unstable" because it is non-deterministic, then indeed sorting of some kind is necessary unless we workaround somehow, and we should look into optimizing that somehow accounting on the fact that we might just need to discover that something changed instead of knowing the ultimate order.
again, worth looking in more detail on how git does it...